### PR TITLE
List of Products Story - GET /api/wines

### PR DESCRIPTION
**Pull request by skp96**
[Link to Story](https://trello.com/c/74FMq4x6/114-list-of-products6)

Highlights
- Added a Products context and create a wines schema, along with tests
- Added an API endpoint; `GET /api/wines` which is handled by the `WineController` `index` action (please see `lib/wine_o_inventory_api_web/controllers/api/wine_controller.ex`)
- Tested the `WineController` `index` action (please see `test/wine_o_inventory_api_web/controllers/wine_controller_tests.exs`)
- Created and tested `WineView` to render data (please see `lib/wine_o_inventory_api_web/views/api/wine_view.ex` and `test/wine_o_inventory_api_web/views/api/wine_view_test.exs`)
- Seeded the database with initial wine data (please see `priv/repo/seeds.exs`)
- Enabled CORS using `cors_plug` (please see `lib/wine_o_inventory_api_web/endpoint.ex`)

### CODE REVIEW

File: lib/wine_o_inventory_api/products.ex [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  @doc """
+  Returns the list of wines.
+
+  ## Examples
+
+      iex> list_wines()
+      [%Wine{}, ...]
+
+  """
+  def list_wines do
```
Comment (JeromeGoodrich) [inline lib/wine_o_inventory_api/products.ex:18]: Seems like for this story this is really all we need?

File: lib/wine_o_inventory_api_web/endpoint.ex [hunk 2]
```diff
@@ -44,5 +44,6 @@ defmodule WineOInventoryApiWeb.Endpoint do
   plug Plug.MethodOverride
   plug Plug.Head
   plug Plug.Session, @session_options
+  plug CORSPlug, origin: "*"
```
Comment (JeromeGoodrich) [inline lib/wine_o_inventory_api_web/endpoint.ex:47]: Why was this needed?

File: priv/repo/migrations/20220121222855_create_wines.exs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+defmodule WineOInventoryApi.Repo.Migrations.CreateWines do
+  use Ecto.Migration
```
Comment (JeromeGoodrich) [inline priv/repo/migrations/20220121222855_create_wines.exs:2]: I'd usually expect a migration to be its own PR

Comment (JeromeGoodrich) [reply priv/repo/migrations/20220121222855_create_wines.exs:2]: Why do you think that is?

File: priv/repo/migrations/20220121222855_create_wines.exs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+  def change do
+    create table(:wines) do
+      add :name, :string, null: false
+      add :description, :string
+      add :rating, :integer
```
Comment (k3ntako) [inline priv/repo/migrations/20220121222855_create_wines.exs:8]: do you think ratings could have decimal points?

File: lib/wine_o_inventory_api_web/views/api/wine_view.ex [hunk 5]
```diff
@@ -0,0 +1,17 @@
+defmodule WineOInventoryApiWeb.WineView do
+    use WineOInventoryApiWeb, :view
+
+    def render("index.json", %{wines: wines}) do
+        render_many(wines, WineOInventoryApiWeb.WineView, "wine.json")
+    end
+
+    def render("wine.json", %{wine: wine}) do
```
Comment (JeromeGoodrich) [inline lib/wine_o_inventory_api_web/views/api/wine_view.ex:8]: What does this do for getting a list of wines?

File: test/wine_o_inventory_api/products_test.exs [hunk 6]
```diff
@@ -0,0 +1,65 @@
+defmodule WineOInventoryApi.ProductsTest do
+  use WineOInventoryApi.DataCase
+
+  alias WineOInventoryApi.Products
+
+  describe "wines" do
+    alias WineOInventoryApi.Products.Wine
+
+    import WineOInventoryApi.ProductsFixtures
```
Comment (JeromeGoodrich) [inline test/wine_o_inventory_api/products_test.exs]: seems a little weird to have these within a describe block. Is that a new Elixir thing?

File: priv/repo/migrations/20220121222855_create_wines.exs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+defmodule WineOInventoryApi.Repo.Migrations.CreateWines do
+  use Ecto.Migration
+
+  def change do
+    create table(:wines) do
+      add :name, :string, null: false
```
Comment (JeromeGoodrich) [inline priv/repo/migrations/20220121222855_create_wines.exs:6]: Should this be unique? Or is it ok to have different wines with the same name?

File: test/support/fixtures/products_fixtures.ex [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  def wine_fixture(attrs \\ %{}) do
+    {:ok, wine} =
+      attrs
+      |> Enum.into(%{
+        description: "some description",
+        name: "some name",
+        quantity: 42,
+        rating: 5
+      })
+      |> WineOInventoryApi.Products.create_wine()
```
Comment (JeromeGoodrich) [inline test/support/fixtures/products_fixtures.ex]: Be careful with fixtures it's east to become overly reliant on them. for the tests like the ones you've written so far, I'd hold off on creating a fixture until I really see the need for it.

**Review by JeromeGoodrich — APPROVED** [review]
Overall, this PR is good, but it's doing too much, it provides all CRUD actions instead of just listing wines. In general, for this sort of backend work, I like to split my work up very small. First I provide a migration, then I make sure I can create the entity, then I try and retrieve it.

File: priv/repo/migrations/20220121222855_create_wines.exs (hunk 4 above)
Comment (k3ntako) [reply priv/repo/migrations/20220121222855_create_wines.exs:8]: Also, does `create table` automatically add an ID? Do you know what type it is (e.g., integer or UUID)?

File: lib/wine_o_inventory_api/products.ex (hunk 1 above)
Comment (skp96) [reply lib/wine_o_inventory_api/products.ex:18]: Yes. I ended up using phoenix's context generator, which generated the context, schema for `Wine`, and all the tests.

File: lib/wine_o_inventory_api_web/endpoint.ex (hunk 2 above)
Comment (skp96) [reply lib/wine_o_inventory_api_web/endpoint.ex:47]: While working on the frontend, my requests weren't able to hit the endpoint due to CORS not being enabled on the backend. Phoenix's CORSPlug allows me to enable cross resource sharing. I should be more specific and not use the wildcard for origin now that I think about it. Thoughts?

File: priv/repo/migrations/20220121222855_create_wines.exs (hunk 3 above)
Comment (skp96) [reply priv/repo/migrations/20220121222855_create_wines.exs:2]: I'm taking an educated guess here. I'm thinking it has to do with ensuring the migrations work correctly, before including code that interacts with the database?  Also, when working on a team, including migrations and code that interacts with a database that doesn't work correctly can cause problems if merged into master?

File: lib/wine_o_inventory_api_web/views/api/wine_view.ex (hunk 5 above)
Comment (skp96) [reply lib/wine_o_inventory_api_web/views/api/wine_view.ex:8]: From an example on the docs, it seems that under the hood, it's following something similar to `Enum.map`. `render_many` maps over each item in `wines` and calls `WineView.render("wine.json", %{wine: wine})`, which returns a `Map`. The result is a `List` of `Maps`

File: test/wine_o_inventory_api/products_test.exs (hunk 6 above)
Comment (skp96) [reply test/wine_o_inventory_api/products_test.exs]: Not sure if it's a new thing. While I was reading through the documentation on testing in phoenix, they utilize `describe` blocks to group similar tests together. So for example, if we defined a `Soda` schema, tests related to that resource would be grouped together inside a `describe` block within `products_test.exs`

File: priv/repo/migrations/20220121222855_create_wines.exs (hunk 7 above)
Comment (skp96) [reply priv/repo/migrations/20220121222855_create_wines.exs:6]: Debatable! Typically wines can have the same name because they are named after the vineyard/company but are made from different wine grapes. For example, you might have a Josh Cellar Pinot Noir and a Jose Cellar Merlot. I was planning on adding a category section later to differentiate the different wines.

File: test/support/fixtures/products_fixtures.ex (hunk 8 above)
Comment (skp96) [reply test/support/fixtures/products_fixtures.ex]: This fixture was provided to me when I ran the context generator test. It came in handy in the `wine_view_test.exs` file.

File: lib/wine_o_inventory_api/products.ex (hunk 1 above)
Comment (cdemyanovich) [reply lib/wine_o_inventory_api/products.ex:18]: So you're going to remove all of the code not related to getting a list of wines, correct?

File: lib/wine_o_inventory_api/products.ex [hunk 9]
```diff
@@ -0,0 +1,104 @@
+defmodule WineOInventoryApi.Products do
+  @moduledoc """
+  The Products context.
+  """
+
+  import Ecto.Query, warn: false
```
Comment (cdemyanovich) [inline lib/wine_o_inventory_api/products.ex]: What does `warn: false` mean/do?

File: lib/wine_o_inventory_api/products.ex (hunk 1 above)
Comment (skp96) [reply lib/wine_o_inventory_api/products.ex:18]: Yes, I'll go ahead and remove the code not related to getting a list of wines and remove tests not related to getting a list of wines as well.

File: priv/repo/migrations/20220121222855_create_wines.exs (hunk 4 above)
Comment (skp96) [reply priv/repo/migrations/20220121222855_create_wines.exs:8]: Yes ratings could have decimal points. If required, I can run a migration to change it. By default, a schema generates a unique ID that has a type of `integer`

File: lib/wine_o_inventory_api/products.ex (hunk 9 above)
Comment (skp96) [reply lib/wine_o_inventory_api/products.ex]: This was generated by the context generator, and I can't seem to find documentation on `warn: false` when importing `Ecto.Query`, but from reading this question on the [elixir form](https://elixirforum.com/t/ecto-query-warning-on-compilation/5938), it seems that its a flag that suppresses warnings that `Ecto.Query` might throw when performing a query?

Comment (cdemyanovich) [reply lib/wine_o_inventory_api/products.ex]: Suppressing warnings is a red flag for me. A program should compile and run without warnings. A website should load without warnings/errors in the browser console. I'd be curious to know if you receive any warnings if you remove `warn: false`, and, if so, what are they?

File: lib/wine_o_inventory_api_web/endpoint.ex (hunk 2 above)
Comment (JeromeGoodrich) [reply lib/wine_o_inventory_api_web/endpoint.ex:47]: Thanks for asking this question if forced me to refresh my CORS knowledge. CORS is essentially away to relax the [same origin policy](https://en.wikipedia.org/wiki/Same-origin_policy#Origin_determination_rules). The same origin policy determination is same scheme, host and port. So when your server is on http://localhost:3000 and your frontend is on http://localhost:8081 you have the same scheme, the same host, but not the same port, therefore you aren't able to access the localhost:3000 from localhost:8081 — they don't comply with the same origin policy. You can get around this with CORS, which is a relaxation of this policy. CORS allows you to specify, who or what can access your server's endpoint. As such, you need to be thoughtful about when you enable CORS. This article provides a pretty good overview of the security risks: https://nodeployfriday.com/posts/cors-cyber-attacks/

File: priv/repo/migrations/20220121222855_create_wines.exs (hunk 3 above)
Comment (JeromeGoodrich) [reply priv/repo/migrations/20220121222855_create_wines.exs:2]: Ding Ding Ding 🔔

File: lib/wine_o_inventory_api_web/views/api/wine_view.ex (hunk 5 above)
Comment (JeromeGoodrich) [reply lib/wine_o_inventory_api_web/views/api/wine_view.ex:8]: So you need both methods?

File: test/wine_o_inventory_api/products_test.exs (hunk 6 above)
Comment (JeromeGoodrich) [reply test/wine_o_inventory_api/products_test.exs]: I meant more specifically the import statements.

File: priv/repo/migrations/20220121222855_create_wines.exs (hunk 4 above)
Comment (JeromeGoodrich) [reply priv/repo/migrations/20220121222855_create_wines.exs:8]: Would you run a new migration or change this one since it hasn't been merged yet?
